transport/src/main/java/io/netty/bootstrap/ChannelInitializerExtension.java

```diff
@@ -0,0 +1,121 @@
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.bootstrap;
+
+import io.netty.channel.Channel;
+import io.netty.channel.ServerChannel;
+
+/**
+ * A channel initializer extension make it possible to enforce rules and apply modifications across multiple,
+ * disconnected uses of Netty within the same JVM process.
+ * <p>
+ * For instance, application-level firewall rules can be injected into all uses of Netty within an application,
+ * without making changes to such uses that are otherwise outside the purview of the application code,
+ * such as 3rd-party libraries.
+ * <p>
+ * All channel initializer extensions that are available on the classpath will be
+ * {@linkplain java.util.ServiceLoader#load(Class) service-loaded} and used by all {@link AbstractBootstrap} subclasses,
+ * unless the extensions are explicitly disabled on the given bootstrap instance with a call to
+ * {@link AbstractBootstrap#disableChannelInitializerExtensions()}.
+ */
+public abstract class ChannelInitializerExtension {
+    /**
+     * Inspect the information in the given {@link ApplicableInfo} object, and determine if this extension is applicable
+     * for the given context.
+     * <p>
+     * If the extension is not applicable, then it won't be called.
+     * <p>
+     * This method may be called multiple times with different parameters for different contexts.
+     *
+     * @param info An object carrying information about the context of where and how this extension would be used.
+     * @return {@code true} if this extension is interested in getting called in the given context,
+     * otherwise {@code false}.
+     */
+    public abstract boolean isApplicable(ApplicableInfo info);
+
+    /**
+     * Get the "priority" of this extension. If multiple extensions are
+     * {@linkplain #isApplicable(ApplicableInfo) applicable} to a given context, then they will be called in their
+     * priority order, from lowest to highest.
+     * <p>
+     * Implementers are encouraged to pick a number between {@code -100.0} and {@code 100.0}, where extensions that have
+     * no particular opinion on their priority are encouraged to return {@code 0.0}.
+     * <p>
+     * Extensions with lower priority will get called first, while extensions with greater priority may be able to
+     * observe the effects of extensions with lesser priority.
+     *
+     * @return The priority.
+     */
+    public double priority() {
```

Comment:
1. Curious why `double` instead of `int/long`? To allow users sneaking in any position, even if other extensions use `0` and `1.0`?
2. Usually, the recommended approach for `ServiceLoader` providers to implement them in a way that order doesn't matter. Even if we add `priority`, users can easily abuse it by always specifying highest/lowest priority for their extensions, resulting in the same undefined ordering as `ServiceLoader` provides by default. Consider adding it later, if there is a use-case. As we learn how users use extensions, there might be a better approach instead of a double. For example, in the future we can add smth like `before(Class)`/`after(Class)` if there is a need to execute one extension before/after another.

Reply:
> Usually, the recommended approach for ServiceLoader providers to implement them in a way that order doesn't matter.

That is still the recommended approach, hence a default implementation is provided, returning a default value.

> For example, in the future we can add smth like `before(Class)`/`after(Class)` if there is a need to execute one extension before/after another.

Such methods can create ambiguous partial orderings. A value with inherent total ordering cannot.

Reply:
Don't have a strong opinion here, we can defer adding priorities or keep it as-is.

> Such methods can create ambiguous partial orderings. A value with inherent total ordering cannot.

The total ordering requires developers to know what all other providers exist and even after they assign a priority value for their provider, it doesn't mean that the other provider can not change its priority later. With before/after it's easier to target explicit providers. For example, a security provider can always be after logging provider. On the other hand, before/after required explicit dependency on another provider to access its class. So, each approach has pros and cons.